python/hidet/ir/Grammar.md

```diff
@@ -0,0 +1,101 @@
+### Notes
+Terminals: `t`
+Non-terminals: t
+
+t?: one or more
+t*: zero or more
+
+ALL_CAPS: Predefined terminal
+- STRING: A string enclosed by `""`
+- IDENT: A valid identifier in C/Python
+- INT/FLOAT/STRING: A valid literal in Python
+
+$\epsilon$ : Null
+( t | $\epsilon$ ) : Optional
+
+#### Minutiae
+- `@` to prevent ambiguity with Call
+- Every statement has an optional `#` IDENT, which encodes an optional attribute dict, the same attribute dict is defined in the top level module space
+- We encode things that are not part of the grammar as an attribute, for example, layouts. This keeps everything relatively extensible
+
+## Grammar
+Value := STRING | INT | FLOAT | Dict | List | Tuple
+Pair := STRING `:` Value
+Tuple := `(` ( Value | $\epsilon$ ) (`,` Value)\* `)`
+List := `[` ( Value | $\epsilon$ ) (`,` Value)\* `]`
+Dict := `{` ( Pair | $\epsilon$ ) (`,` Pair)\* `}`
+
+AttributeName := `#` IDENT
+Attribute := AttributeName `=` Dict `;`
+
+DataType := `i1` | `i8` | `i16` | `i32` | `i64` | `u8` | `u16` | `u32` | `u64` | `f16` | `f32` | `f64` | `void` | `str`
+PtrType := `~`DataType
+TensorType := DataType`<` NAT (`,` NAT)\* (`,` AttributeName | $\epsilon$ ) `>`
+TensorPtrType := `~`TensorType
+Type := DataType | PtrType | TensorType | TensorPtrType
+
+Module := `{` ( Function | Attribute )\* `}` 
+Function := `def` IDENT `(` ( Arg`:` Type | $\epsilon$ ) (`,` Arg`:` Type)\*` )` `->` Type `{` Stmt* `}`
+Stmt := CoreStmt ( AttributeName `;` | $\epsilon$ )
+CoreStmt := DeclareStmt 
+		| BufferStoreStmt 
+		| AssignStmt 
+		| ReturnStmt 
+		| LetStmt 
+		| ForStmt 
+		| WhileStmt
+		| BreakStmt
+		| ContinueStmt
+		| IfStmt
+		| AssertStmt
+		| BlackBoxStmt
+		| AsmStmt
+
+DeclareStmt := `decl` IDENT `:` Type ( `=` Expr | $\epsilon$ ) `;`
+BufferStoreStmt := IDENT `[`Expr`]` `=` Expr `;`
+AssignStmt := IDENT `=` Expr `;`
+ReturnStmt := `return` Expr `;`
+LetStmt := `let` IDENT `:` Type `=` Expr  ( `in` Stmt | $\epsilon$ ) `;`
+ForStmt := `for` ( IDENT | `(` IDENT (`,` IDENT)\* `)` ) `in` Mapping `{` Stmt* `}`
+WhileStmt := `while` Expr `{` Stmt* `}`
+BreakStmt := `break` `;`
+ContinueStmt := `continue` `;`
+IfStmt := `if` Expr `{` Stmt* `}` ( `else` `{` Stmt* `}` | $\epsilon$ )
+AssertStmt := `assert` Expr `;`
+BlackBoxStmt := `BlackBox` `{` STRING `}`
+AsmStmt := ( `volatile` | $\epsilon$ ) `asm` `{ in {` IDENT (`,` IDENT) `}` `out` `{` IDENT (`,` IDENT) `}` `{` STRING `}` `}`
+
+
+BinaryOp := `+` | `*` | `-` | `/` | `//` | `%` | `**` | `^` | `<<` | `>>` | `||` | `&&` | `==` | `!=` | `<` | `>` | `<=` | `>=`
+UnaryOp := `-` | `!` 
+Expr := `(` Expr BinaryOp Expr `)`
+		| `(` UnaryOp Expr `)`
```

Comment:
Better not add the () here.
Use something like
```
Expr := Expr BinaryOp Expr
  | UnaryOp Expr
  | `(` Expr `)`
```

Reply:
Unfortunately, this makes the grammar ambiguous, so it would take some non-trivial restructing to preserve order of operations and make the grammar unambiguous.

Reply:
I see. 
When using antlr4, we can use the order of the derivatives to specify their priority, like [here](https://github.com/yaoyaoding/csc488/blob/a273aae43f1f774c23f833f6d7ace5edb61dce54/assignments/a2/minicc/grammars/MiniC.g4#L70-L86).

Reply:
Not sure whehter the parser generator we are gonna to use support this feature.

Reply:
Can I know whether the lark support that feature like antlr4?

I have see the new grammar about expression, and I feel it is not as readable as the one that use order to determine priority. And it might increase the difficulty to convert the AST to hidet ir if you define it in this way. But I will leave it to you to decide (or you can come back to update the grammar when you implement the conversion).

Reply:
No, I don't believe that it is possible in lark.
Alright, I am working on the IR dumper right now, since by default it dumps all binary and unary expressions in brackets, I am not sure if it is necessary. Since to remove the reliance of always printing brackets I would have to inspect the parse tree ahead of time to determine the order of operations.

Reply:
In ANTLR4, we can define the grammar like
```
BINARY_EXPR := EXPR * EXPR
            |= EXPR + EXPR
```
Then, the parser will generate AST for "a+b*c" as <ADD(a, MUL(b, c))> and "a * b + c" as <ADD(MUL(a, b), c)>. 

We better specify the operator priority in the grammar, instead of analying after transformation (it that's possible). 

Also, it is not a good idea to force the text form to always have "()" like "(a+(b+((e+f)*d)))".

Reply:
The python grammar implemented in lark: https://github.com/lark-parser/lark/blob/master/examples/advanced/python2.lark

Reply:
If you find lark is not powerful enough, we can also consider using antlr4: https://alanhohn.com/posts/2016/antlr-python/
(if we use antlr4, we can consider put both the grammar and the generated python source code in our git repo).